### PR TITLE
Plugin loader error handling tweaks

### PR DESCRIPTION
**Pull request by kaixi-wang**
## What changes are proposed in this pull request?

Prevent attempting to registering and loading a plugin that has errors
image

## How is this patch tested? If it is not, please explain why.

(Details)

### What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [ ] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [ ] Other

### CODE REVIEW

File: fiftyone/operators/loader.py [hunk 1]
```diff
@@ -89,11 +92,13 @@ def register_module(plugin_definition, mod):
     """
     try:
         pctx = PluginContext(plugin_definition)
-        KNOWN_PLUGIN_CONTEXTS[pctx.name] = pctx
+        if pctx.errors:
+            # Don't register if there are errors
+            return pctx
         mod.register(pctx)
```
Comment (ritch) [inline fiftyone/operators/loader.py:98]: This should be called even with errors

File: fiftyone/operators/loader.py [hunk 2]
```diff
@@ -89,11 +92,13 @@ def register_module(plugin_definition, mod):
     """
     try:
         pctx = PluginContext(plugin_definition)
-        KNOWN_PLUGIN_CONTEXTS[pctx.name] = pctx
+        if pctx.errors:
```
Comment (ritch) [inline fiftyone/operators/loader.py:95]: This isn't needed because we actually want to register the errors.

File: fiftyone/operators/loader.py [hunk 3]
```diff
@@ -157,6 +162,15 @@ def exec_module_from_dir(module_dir, plugin_definition):
         )
     spec = importlib.util.spec_from_file_location(mod_dir, mod_filepath)
     mod = importlib.util.module_from_spec(spec)
+    try:
+        pctx = register_module(plugin_definition, mod)
```
Comment (ritch) [inline fiftyone/operators/loader.py:166]: We don't want to call `register_module()` twice.

File: fiftyone/operators/loader.py [hunk 4]
```diff
@@ -157,6 +162,15 @@ def exec_module_from_dir(module_dir, plugin_definition):
...
+    except Exception as e:
+        raise e
+    if pctx.has_errors():
+        raise ValueError(
+            f"Plugin {plugin_definition.name} has errors: {pctx.errors}"
+        )
+    # only add to sys.modules if there are no errors
```
Comment (ritch) [inline fiftyone/operators/loader.py:173]: This seems like the only changed needed - to only add to sys.modules if there are not errors, although I'm not sure you can attempt to load `exec_module` without it added, I've not tried that.

File: fiftyone/operators/loader.py (hunk 1 above)
Comment (kaixi-wang) [reply fiftyone/operators/loader.py:96]: updated and put in `finally` so whether or not an error is raised, it gets added

File: fiftyone/operators/loader.py (hunk 3 above)
Comment (kaixi-wang) [reply fiftyone/operators/loader.py:171]: yeah I forgot to remove the previous call. Fixed now

File: fiftyone/operators/loader.py [hunk 5]
```diff
@@ -87,13 +90,21 @@ def register_module(plugin_definition, mod):
...
     except Exception as e:
-        errors = [traceback.format_exc()]
-        pctx.errors = errors
+        if pctx:
+            pctx.errors.append(traceback.format_exc())
+    finally:
+        if pctx and pctx.name:
+            KNOWN_PLUGIN_CONTEXTS[pctx.name] = pctx
+        else:
+            # If plugin context is None, then there is no plugin name to
```
Comment (ritch) [inline fiftyone/operators/loader.py:104]: We should print the plugin_definition.directory

Comment (kaixi-wang) [reply fiftyone/operators/loader.py:107]: added logging

**Pull request closed without merging**